File: archiso/airootfs/usr/share/churros/preferences/services/dotfiles/foot_config.py
import os
import tempfile


class FootConfig:

    PATH = os.path.join(
        os.path.expanduser("~"),
        ".config",
        "foot",
        "foot.ini"
    )

    @classmethod
    def _read(cls):

        try:

            with open(
                cls.PATH,
                "r"
            ) as file:

                return file.readlines()

        except Exception:

            return []

    @classmethod
    def _write_atomic(
        cls,
        lines
    ):

        directory = os.path.dirname(
            cls.PATH
        )

        os.makedirs(
            directory,
            exist_ok=True
        )

        fd, tmp = tempfile.mkstemp(
            prefix="foot-",
            suffix=".ini",
            dir=directory
        )

        try:

            with os.fdopen(
                fd,
                "w"
            ) as file:

                file.writelines(
                    lines
                )

            os.replace(
                tmp,
                cls.PATH
            )

        except Exception:

            try:

                os.unlink(
                    tmp
                )

            except OSError:

                pass

            raise

    @classmethod
    def _find_section(
        cls,
        lines,
        section
    ):

        for i, line in enumerate(
            lines
        ):

            stripped = line.strip()

            if stripped == "[" + section + "]":

                return i

        return -1

    @classmethod
    def _set_key(
        cls,
        lines,
        section,
        key,
        value
    ):

        section_idx = cls._find_section(
            lines,
            section
        )

        if section_idx == -1:

            lines.append(
                "\n[" + section + "]\n"
            )

            section_idx = len(lines) - 1

        end = len(lines)

        for j in range(
            section_idx + 1,
            len(lines)
        ):

            stripped = lines[j].strip()

            if stripped.startswith(
                "["
            ) and stripped.endswith(
                "]"
            ):

                end = j

                break

        for j in range(
            section_idx + 1,
            end
        ):

            stripped = lines[j].strip()

            if stripped.startswith(
                key + "="
            ) or stripped.startswith(
                key + " "
            ) or stripped == key + "=":

                prefix = lines[j][:len(lines[j]) - len(lines[j].lstrip())]

                lines[j] = prefix + key + "=" + str(
                    value
                ) + "\n"

                return True

        insert = "    " + key + "=" + str(
            value
        ) + "\n"

        lines.insert(
            end,
            insert
        )

        return True

    @classmethod
    def set_font(
        cls,
        font
    ):

        lines = cls._read()

        cls._set_key(
            lines,
            "main",
            "font",
            font
        )

        cls._write_atomic(
            lines
        )

    @classmethod
    def set_dark(
        cls,
        dark
    ):

        lines = cls._read()

        target_section = "colors-dark" if dark else "colors-light"

        existing_dark = cls._find_section(
            lines,
            "colors-dark"
        ) != -1

        existing_light = cls._find_section(
            lines,
            "colors-light"
        ) != -1

        section_idx = cls._find_section(
            lines,
            target_section
        )

        if section_idx == -1:

            if dark and not existing_dark:

                lines.append(
                    "[colors-dark]\n"
                )

            elif not dark and not existing_light:

                lines.append(
                    "[colors-light]\n"
                )

        cls._write_atomic(
            lines
        )

    # ------------------------------------------------------------ Getters

    @classmethod
    def _get_key(cls, section, key, default=""):

        lines = cls._read()
        idx = cls._find_section(lines, section)
        if idx < 0:
            return default

        for j in range(idx + 1, len(lines)):
            stripped = lines[j].strip()
            if stripped.startswith("[") and stripped.endswith("]"):
                break
            if stripped.startswith(key + "="):
                return stripped.split("=", 1)[1].strip()

        return default

    @classmethod
    def get_font(cls):

        return cls._get_key("main", "font", "JetBrainsMono Nerd Font:size=10")

    @classmethod
    def get_pad(cls):

        return cls._get_key("main", "pad", "8x8")

    @classmethod
    def get_cursor_style(cls):

        return cls._get_key("cursor", "style", "beam")

    @classmethod
    def get_cursor_blink(cls):

        v = cls._get_key("cursor", "blink", "yes")
        return v.lower() in ("yes", "true", "1")

    @classmethod
    def get_bell(cls):

        v = cls._get_key("bell", "urgent", "yes")
        return v.lower() in ("yes", "true", "1")

    @classmethod
    def get_hide_when_typing(cls):

        v = cls._get_key("mouse", "hide-when-typing", "yes")
        return v.lower() in ("yes", "true", "1")

    # ------------------------------------------------------------- Setters

    @classmethod
    def set_pad(cls, pad):

        lines = cls._read()
        cls._set_key(lines, "main", "pad", pad)
        cls._write_atomic(lines)

    @classmethod
    def set_cursor(cls, style, blink):

        lines = cls._read()
        cls._set_key(lines, "cursor", "style", style)
        cls._set_key(lines, "cursor", "blink", "yes" if blink else "no")
        cls._write_atomic(lines)

    @classmethod
    def set_bell(cls, urgent):

        lines = cls._read()
        cls._set_key(lines, "bell", "urgent", "yes" if urgent else "no")
        cls._set_key(lines, "bell", "notify", "yes" if urgent else "no")
        cls._write_atomic(lines)

    @classmethod
    def set_hide_when_typing(cls, hide):

        lines = cls._read()
        cls._set_key(lines, "mouse", "hide-when-typing", "yes" if hide else "no")
        cls._write_atomic(lines)

    @classmethod
    def reload(cls):

        try:
            import subprocess
            subprocess.Popen(
                ["pkill", "-SIGUSR1", "foot"],
                stdout=subprocess.DEVNULL,
                stderr=subprocess.DEVNULL
            )
        except Exception:
            pass

File: rust/preferences/src/services/foot_config.rs
// ==========================================
// FootConfig — ~/.config/foot/foot.ini
// (equivalente a services/dotfiles/foot_config.py)
// ==========================================

use std::fs;
use std::path::PathBuf;
use std::process::Command;

pub struct FootConfig;

fn config_path() -> PathBuf {
    let home = std::env::var("HOME").unwrap_or_else(|_| "/root".to_string());
    PathBuf::from(home).join(".config").join("foot").join("foot.ini")
}

fn read_lines() -> Vec<String> {
    match fs::read_to_string(config_path()) {
        Ok(content) => content.lines().map(|s| s.to_string()).collect(),
        Err(_) => Vec::new(),
    }
}

fn write_atomic(lines: &[String]) {
    let path = config_path();
    if let Some(parent) = path.parent() {
        let _ = fs::create_dir_all(parent);
    }
    // Escritura atómica: tmp + rename
    let tmp = path.with_extension("ini.tmp");
    let content = lines.join("\n");
    if fs::write(&tmp, content).is_ok() {
        let _ = fs::rename(&tmp, &path);
    }
}

fn find_section(lines: &[String], section: &str) -> isize {
    for (i, line) in lines.iter().enumerate() {
        if line.trim() == format!("[{section}]") {
            return i as isize;
        }
    }
    -1
}

/// Actualiza o inserta `key=value` dentro de la sección (equivalente a _set_key).
fn set_key(lines: &mut Vec<String>, section: &str, key: &str, value: &str) {
    let mut section_idx = find_section(lines, section);
    if section_idx == -1 {
        lines.push(format!("\n[{section}]\n"));
        section_idx = (lines.len() - 1) as isize;
    }

    let mut end = lines.len();
    for j in (section_idx as usize + 1)..lines.len() {
        let stripped = lines[j].trim();
        if stripped.starts_with('[') && stripped.ends_with(']') {
            end = j;
            break;
        }
    }

    for j in (section_idx as usize + 1)..end {
        let stripped = lines[j].trim();
        if stripped.starts_with(&format!("{key}=")) || stripped.starts_with(&format!("{key} ")) {
            let prefix: String = lines[j]
                .chars()
                .take_while(|c| *c == ' ' || *c == '\t')
                .collect();
            lines[j] = format!("{prefix}{key}={value}");
            return;
        }
    }

    let insert = format!("    {key}={value}");
    lines.insert(end, insert);
}

fn get_key(section: &str, key: &str, default: &str) -> String {
    let lines = read_lines();
    let idx = find_section(&lines, section);
    if idx < 0 {
        return default.to_string();
    }
    for j in (idx as usize + 1)..lines.len() {
        let stripped = lines[j].trim();
        if stripped.starts_with('[') && stripped.ends_with(']') {
            break;
        }
        if let Some(rest) = stripped.strip_prefix(&format!("{key}=")) {
            return rest.trim().to_string();
        }
    }
    default.to_string()
}

fn get_key_bool(section: &str, key: &str, default: bool) -> bool {
    let v = get_key(section, key, if default { "yes" } else { "no" });
    let v = v.to_lowercase();
    v == "yes" || v == "true" || v == "1"
}

impl FootConfig {
    // ------------------------------------------------------------ Getters

    pub fn get_font() -> String {
        get_key("main", "font", "JetBrainsMono Nerd Font:size=10")
    }

    pub fn get_pad() -> String {
        get_key("main", "pad", "8x8")
    }

    pub fn get_cursor_style() -> String {
        get_key("cursor", "style", "beam")
    }

    pub fn get_cursor_blink() -> bool {
        get_key_bool("cursor", "blink", true)
    }

    pub fn get_bell() -> bool {
        get_key_bool("bell", "urgent", true)
    }

    pub fn get_hide_when_typing() -> bool {
        get_key_bool("mouse", "hide-when-typing", true)
    }

    // ------------------------------------------------------------- Setters

    pub fn set_font(font: &str) {
        let mut lines = read_lines();
        set_key(&mut lines, "main", "font", font);
        write_atomic(&lines);
    }

    /// Solo crea la sección [colors-dark]/[colors-light] si falta (como set_dark).
    #[allow(dead_code)] // portado por paridad; sin uso en las páginas actuales
    pub fn set_dark(dark: bool) {
        let mut lines = read_lines();
        let target_section = if dark { "colors-dark" } else { "colors-light" };
        let existing_dark = find_section(&lines, "colors-dark") != -1;
        let existing_light = find_section(&lines, "colors-light") != -1;

        if find_section(&lines, target_section) == -1 {
            if dark && !existing_dark {
                lines.push("[colors-dark]\n".to_string());
            } else if !dark && !existing_light {
                lines.push("[colors-light]\n".to_string());
            }
        }
        write_atomic(&lines);
    }

    pub fn set_pad(pad: &str) {
        let mut lines = read_lines();
        set_key(&mut lines, "main", "pad", pad);
        write_atomic(&lines);
    }

    pub fn set_cursor(style: &str, blink: bool) {
        let mut lines = read_lines();
        set_key(&mut lines, "cursor", "style", style);
        set_key(&mut lines, "cursor", "blink", if blink { "yes" } else { "no" });
        write_atomic(&lines);
    }

    pub fn set_bell(urgent: bool) {
        let mut lines = read_lines();
        let v = if urgent { "yes" } else { "no" };
        set_key(&mut lines, "bell", "urgent", v);
        set_key(&mut lines, "bell", "notify", v);
        write_atomic(&lines);
    }

    pub fn set_hide_when_typing(hide: bool) {
        let mut lines = read_lines();
        set_key(
            &mut lines,
            "mouse",
            "hide-when-typing",
            if hide { "yes" } else { "no" },
        );
        write_atomic(&lines);
    }

    /// pkill -SIGUSR1 foot (recarga la config en las terminales abiertas)
    pub fn reload() {
        let _ = Command::new("pkill")
            .args(["-SIGUSR1", "foot"])
            .stdout(std::process::Stdio::null())
            .stderr(std::process::Stdio::null())
            .spawn();
    }
}
